README.md
# metalsmith-github-profile

[![npm Version](https://badgen.net/npm/v/metalsmith-github-profile?icon=npm)](https://www.npmjs.com/package/metalsmith-github-profile)
[![npm Weekly Downloads](https://badgen.net/npm/dw/metalsmith-github-profile)](https://www.npmjs.com/package/metalsmith-github-profile)

[![Known Vulnerabilities](https://snyk.io/test/npm/metalsmith-github-profile/badge.svg)](https://snyk.io/test/npm/metalsmith-github-profile)
[![Test Coverage](https://badgen.net/codecov/c/github/emmercm/metalsmith-github-profile/main?icon=codecov)](https://codecov.io/gh/emmercm/metalsmith-github-profile)
[![Maintainability](https://badgen.net/codeclimate/maintainability/emmercm/metalsmith-github-profile?icon=codeclimate)](https://codeclimate.com/github/emmercm/metalsmith-github-profile/maintainability)

[![GitHub](https://badgen.net/badge/emmercm/metalsmith-github-profile/purple?icon=github)](https://github.com/emmercm/metalsmith-github-profile)
[![License](https://badgen.net/github/license/emmercm/metalsmith-github-profile?color=grey)](https://github.com/emmercm/metalsmith-github-profile/blob/main/LICENSE)

A Metalsmith plugin to fetch GitHub profile information as global metadata.

## Installation

```bash
npm install --save metalsmith-github-profile
```

## JavaScript Usage

```javascript
const Metalsmith    = require('metalsmith');
const githubProfile = require('metalsmith-github-profile');

Metalsmith(__dirname)
    .use(githubProfile({
        username: "<your-username-here>"
        // additional options
    }))
    .build((err) => {
        if (err) {
            throw err;
        }
    });
```

## Global metadata

This plugin adds a metadata field named `github.profile` to the global metadata which can be used with templating engines, such as with [`handlebars`](https://www.npmjs.com/package/handlebars):

```handlebars
GitHub username: {{ github.profile.user.login }}

The rest of the page content.
```

The following metadata is available from the GitHub public API:

- [`github.profile.user`](https://docs.github.com/en/rest/reference/users#get-a-user) (`Object`)
- [`github.profile.repos`](https://docs.github.com/en/rest/reference/repos#list-repositories-for-a-user) (`Object[]`)

## Options

### `username` (required)

Type: `string`

The GitHub username to fetch information for.

### `authorization` (optional)

Type: `{username: string, token: string}`

A GitHub username and OAuth token (including personal access tokens) to use for API requests to get around rate limits.

You can source the token from environment variables like this:

```javascript
githubProfile({
  username: "emmercm",
  authorization: {
    username: "emmercm",
    token: process.env.GITHUB_PERSONAL_ACCESS_TOKEN
  }
})
```

### `timeout` (optional)

Type: `number` Default: `5000`

Timeout in milliseconds for API requests to GitHub.

### `retries` (optional)

Type: `number` Default: `3`

Number of times to retry GitHub API requests.

### `retryableStatusCodes` (optional)

Type: `number[]` Default: `[0, 408, 500, 502, 503, 504]`

A list of HTTP status codes to retry GitHub API requests on.

## Changelog

[Changelog](./CHANGELOG.md)


CHANGELOG.md
# Changelog

## v0.2.0 / 2022-08-07

- Fixed HTTP(S) timeouts preventing the plugin from stopping.

## v0.1.0 / 2022-07-28

- Added the `authorization` option.
- Added the rate limit expiration time to rate limit failures.

## v0.0.1 / 2021-11-26

- Initial version.


lib/index.test.js
'use strict';

const {
  existsSync, mkdirSync, readdirSync, readFileSync, statSync,
} = require('fs');
const { join } = require('path');

const Metalsmith = require('metalsmith');
const hbtmd = require('metalsmith-hbt-md');
const handlebars = require('handlebars');
const assertDir = require('assert-dir-equal');

const githubProfile = require('./index');

const test = (dir, config) => {
  describe(dir, () => {
    // Allow src directory to not exist / be empty and not committed
    if (!existsSync(`${dir}/src`)) {
      mkdirSync(`${dir}/src`);
    }

    ['/', '\\'].forEach((pathSeparator) => {
      it(`should build with path separator "${pathSeparator}"`, (testDone) => {
        Metalsmith(`${dir}`)
          // Convert the path separators to the one being tested
          .use((files, metalsmith, done) => {
            Object.keys(files)
              .forEach((filename) => {
                const newFilename = filename.replace(/[/\\]/g, pathSeparator);
                if (newFilename !== filename) {
                  files[newFilename] = files[filename];
                  delete files[filename];
                }
              });
            done();
          })
          // Run the plugin
          .use(githubProfile(config.options))
          .use(hbtmd(handlebars))
          // Convert the path separators back to system default
          .use((files, metalsmith, done) => {
            const properPathSeparator = process.platform === 'win32' ? '\\' : '/';
            Object.keys(files)
              .forEach((filename) => {
                const newFilename = filename.replace(/[/\\]/g, properPathSeparator);
                if (newFilename !== filename) {
                  files[newFilename] = files[filename];
                  delete files[filename];
                }
              });
            done();
          })
          // Test the output
          .build((err) => {
            try {
              if (config.error) {
                expect(err)
                  .toBe(config.error);
              } else {
                expect(err)
                  .toBeNull();
              }

              if (err) {
                testDone();
                return;
              }

              assertDir(`${dir}/build`, `${dir}/expected`, { filter: () => true });
              testDone();
            } catch (assertionError) {
              testDone(assertionError);
            }
          });
      }, 30 * 1000);
    });
  });
};

describe('metalsmith-github-profile', () => {
  const dirs = (p) => readdirSync(p)
    .map((f) => join(p, f))
    .filter((f) => statSync(f).isDirectory());
  dirs('lib/fixtures')
    .forEach((dir) => {
      const config = existsSync(`${dir}/config.json`) ? JSON.parse(readFileSync(`${dir}/config.json`).toString()) : {};
      test(dir, config);
    });
});


lib/index.js
'use strict';

const http = require('http');
const https = require('https');

const async = require('async');
const deepmerge = require('deepmerge');

const retryJsonGet = (link, options, callback, attempt = 1) => {
  const url = new URL(link);
  const library = {
    'http:': http,
    'https:': https,
  }[url.protocol];

  const requestOptions = {
    method: 'GET',
    headers: {
      Accept: 'application/vnd.github.v3+json',
      'User-Agent': 'metalsmith-github-profile',
    },
    timeout: options.timeout,
  };

  let authorized = false;
  if (options.authorization && Object.keys(options.authorization)) {
    const token = `${options.authorization.username}:${options.authorization.token}`;
    requestOptions.headers.Authorization = `Basic ${Buffer.from(token).toString('base64')}`;
    authorized = true;
  }

  const req = library.request(link, requestOptions, (res) => {
    const rateLimitLimit = parseInt(res.headers['x-ratelimit-limit'], 10);
    const rateLimitRemaining = parseInt(res.headers['x-ratelimit-remaining'], 10);
    const rateLimitReset = parseInt(res.headers['x-ratelimit-reset'], 10);

    // Try to warn the user about silent authentication issues if authorization was intended
    if (authorized && rateLimitLimit < 1000) {
      console.log(`WARN: Authorization possibly failed, ${rateLimitRemaining}/${rateLimitLimit} rate limit calls remaining`);
    }

    if (res.statusCode < 200 || res.statusCode > 299) {
      if (options.retryableStatusCodes.includes(res.statusCode)) {
        if (attempt <= options.retries) {
          setTimeout(() => {
            retryJsonGet(link, options, callback, attempt + 1);
          }, (2 ** attempt) * 100);
        } else {
          callback(new Error('retry limit reached'), null);
        }
      } else {
        let message = `HTTP ${res.statusCode}: ${res.statusMessage}`;
        if (rateLimitRemaining <= 0) {
          message += `(rate limit reset at ${(new Date(rateLimitReset * 1000)).toLocaleString()})`;
        }
        callback(new Error(message), null);
      }
      return;
    }

    let data = '';
    res.on('data', (chunk) => {
      data += chunk;
    });
    res.on('end', () => {
      callback(null, JSON.parse(data));
    });
  });

  req.on('error', (err) => {
    callback(err, null);
  });

  req.on('timeout', () => {
    req.destroy(); // cause an error
  });

  req.end();
};

const getUser = (username) => (options, callback) => {
  retryJsonGet(`https://api.github.com/users/${username}`, options, callback);
};

const getRepos = (username, page = 1, results = []) => (options, callback) => {
  retryJsonGet(`https://api.github.com/users/${username}/repos?per_page=100&page=${page}`, options, (err, data) => {
    // Recurse to get all pages
    if (data && data.length) {
      results = [...results, ...data];
      getRepos(username, page + 1, results)(options, callback);
      return;
    }

    // Key the results by repo name
    results = results.reduce((obj, repo) => {
      delete repo.owner; // large, duplicate info
      obj[repo.name] = repo;
      return obj;
    }, {});

    callback(err, results);
  });
};

/**
 * Plugin entrypoint.
 * @param {Object} options
 * @returns {function(Object.<string, Object>, Object, function)}
 */
module.exports = (options) => {
  options = deepmerge({
    timeout: 5 * 1000,
    authorization: {},
    retries: 3,
    retryableStatusCodes: [0, 408, 500, 502, 503, 504],
  }, options || {});

  return (files, metalsmith, done) => {
    const methods = {
      user: getUser(options.username),
      repos: getRepos(options.username),
    };
    async.mapValues(methods, (value, key, callback) => {
      value(options, callback);
    }, (err, result) => {
      if (err) {
        done(err);
        return;
      }

      const metadata = metalsmith.metadata();
      if (!metadata.github) {
        metadata.github = {};
      }
      metadata.github.profile = result;

      done();
    });
  };
};
